node/hyperbahn/service_proxy.js

```diff
@@ -295,11 +299,60 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
 function refreshServicePeer(serviceName, hostPort) {
     var self = this;
 
+    // Add hostPort to service peers, decision pending on whether to connect.
     var peer = self.getServicePeer(serviceName, hostPort);
-    peer.connect();
 
+    // The old way: fully connect every egress to all affine peers.
+    if (!self.partialAffinityEnabled) {
+        peer.connect();
+        return;
+    }
+
+    // Make a note not to collect this service
     var time = self.channel.timers.now();
     self.exitServices[serviceName] = time;
+
+    // Obtain and sort the affine worker and relay lists.
+    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    relays.sort();
+    var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+    var workers = serviceChannel.peers.keys();
+    workers.sort();
+
+    // Find our position within the affine relay set so we can project that
+    // position into the affine worker set.
+    var relayIndex = relays.indexOf(self.channel.hostPort);
```

Comment:
could be binary bisection... #justsayin... not sure where the inflection point is for node indexOf vs bisect